### DISCUSSION TITLE
Toggle icon with ToggleButton onlx in XAML

### DISCUSSION THREAD

**Discussion by StanDeMan** [AvaloniaUI/Avalonia#8005] [Q&A] [answered]
Hi Avalonia lovers,
I'm pretty new to Avalonia and struggling with not available Triggers. How to toggle icons in ToggleButton.
What i have done so far - my xaml:

```
<StackPanel>
		<Border
			Background="Transparent">
			<ToggleButton 
				Width="24"
				Height="24"
				IsChecked="True">
				<ToggleButton.Styles>
					<Style Selector="ToggleSwitch">
						<Setter Property="Opacity" Value="1.0"/>
					</Style>
					<Style Selector="ToggleSwitch">
						<Setter  Property="Opacity" Value="0.0"/>
				   </Style>
				</ToggleButton.Styles>
				<ToggleButton.Template>
					<ControlTemplate>
						<Panel>
							<icons:MaterialIcon
								Width="20"
								Height="20"
								Classes="CollapseAll"
								Kind="ArrowCollapseAll"/>
						</Panel>
					</ControlTemplate>		
				</ToggleButton.Template>
				<ToggleButton.Template>
					<ControlTemplate>
						<Panel>
							<icons:MaterialIcon 
								Width="20"
								Height="20"	
								Classes="ExpandAll"
								Kind="ArrowExpandAll"/>
						</Panel>
					</ControlTemplate>		
				</ToggleButton.Template>
			</ToggleButton>  
		</Border>
	</StackPanel>
</UserControl>
```

This is showing the ArrowExpandAll icon being up in z order. 
What I want to achieve is - if the ToggleButton is pressed - the ArrowCollapseAll icon should be shown - like I comment out the lower part, leting the ArrowCollapseAll be shown:

```
<StackPanel>
		<Border
			Background="Transparent">
			<ToggleButton 
				Width="24"
				Height="24"
				IsChecked="True">
				<ToggleButton.Styles>
					<Style Selector="ToggleSwitch">
						<Setter Property="Opacity" Value="1.0"/>
					</Style>
					<Style Selector="ToggleSwitch">
						<Setter  Property="Opacity" Value="0.0"/>
				   </Style>
				</ToggleButton.Styles>
				<ToggleButton.Template>
					<ControlTemplate>
						<Panel>
							<icons:MaterialIcon
								Width="20"
								Height="20"
								Classes="CollapseAll"
								Kind="ArrowCollapseAll"/>
						</Panel>
					</ControlTemplate>		
				</ToggleButton.Template>
				<!--<ToggleButton.Template>
					<ControlTemplate>
						<Panel>
							<icons:MaterialIcon 
								Width="20"
								Height="20"	
								Classes="ExpandAll"
								Kind="ArrowExpandAll"/>
						</Panel>
					</ControlTemplate>		
				</ToggleButton.Template>-->
			</ToggleButton>  
		</Border>
	</StackPanel>
```

Please help me.
I really do not know how to do this. A simple XAML solution is preferred - having no code behind. 
Everything I tried wasn't working. I am running out of ideas. Thanks for your help. 

Kindest regards,
StanDeMan

**Comment by timunie** [comment, marked as answer]
You could place both icons in the same panel and just bind IsChecked to IsVisible 

```xml
IsVisible="{TemplateBinding IsChecked}"

And the other

IsVisible="{TemplateBinding !IsChecked}"
```

Have written this without teating, so the syntax may be different.

It that's not enough there is also a pseudoclass for this. See the docs about Styling.

Happy coding
Tim

**Reply by StanDeMan** [reply]
Thanks Tim for your rapid answer. I'll give it a try. Yup - Avalonia has this NOT (!) syntax WPF hasn't. Thx StanDeMan

**Reply by StanDeMan** [reply]
Still going crazy. ;)
ArrowExpandAll gets displayed. The next click the ToggleButton is empty - never showing up an icon again?

```
<UserControl 
	xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	xmlns:icons="using:Material.Icons.Avalonia"
    mc:Ignorable="d" 
	d:DesignWidth="24" 
	d:DesignHeight="24"
    x:Class="WhatEver.Gui.UserInterface.Controls.MinMaxToggelButton">
	<StackPanel>
		<Border
			Background="Transparent">
			<ToggleButton 
				Width="24"
				Height="24"
				IsChecked="True">
				<ToggleButton.Template>
					<ControlTemplate>
						<Panel>
							<icons:MaterialIcon
								Width="20"
								Height="20"
								IsVisible="{Binding !IsChecked, RelativeSource={RelativeSource TemplatedParent}, Mode=TwoWay}"
								Kind="ArrowCollapseAll"/>
						</Panel>
					</ControlTemplate>		
				</ToggleButton.Template>
				<ToggleButton.Template>
					<ControlTemplate>
						<Panel>
							<icons:MaterialIcon 
								Width="20"
								Height="20"	
								IsVisible="{Binding IsChecked, RelativeSource={RelativeSource TemplatedParent}, Mode=TwoWay}"
								Kind="ArrowExpandAll"/>
						</Panel>
					</ControlTemplate>		
				</ToggleButton.Template>
			</ToggleButton>  
		</Border>
	</StackPanel>
</UserControl>
```
I tryed with TemplateBinding but it is throwing me an exception at design time?
With the source above something gets triggered. But not as expected. Hm?
Does someone has an running example, please? 
I love Avalonia. The rest is working as expected. But such a simple thing is driving me crazy! :)

Kindest regards - StanDeMan

**Reply by timunie** [reply]
You should not define two templates. Move the other icon in the same Panel as the first one is.

**Reply by timunie** [reply]
A Template can only be defined once per Style.

**Reply by StanDeMan** [reply]
An examples would be nice. Cannot get it running.

Thanks and kindest regards - StanDeMan

**Reply by StanDeMan** [reply]
Thanks a lot for your example. This is how I started a while ago.
Getting this exception - even eith the example:

System.Xaml.XamlException: 'Internal compiler error while transforming node XamlX.Ast.XamlAstObjectNode:
Avalonia.Data.Core.ExpressionParseException: Unexpected '!'.
   at Avalonia.Markup.Xaml.Parsers.PropertyParser.Parse(CharacterReader r) in /_/src/Markup/Avalonia.Markup.Xaml/Parsers/PropertyParser.cs:line 44
   at Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.XamlIlAvaloniaPropertyHelper.CreateNode(AstTransformationContext context, String propertyName, IXamlAstTypeReference selectorTypeReference, IXamlLineInfo lineInfo) in /_/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlAvaloniaPropertyHelper.cs:line 73
   at Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.AvaloniaXamlIlLanguage.CustomValueConverter(AstTransformationContext context, IXamlAstValueNode node, IXamlType type, IXamlAstValueNode& result) in /_/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguage.cs:line 196
   at XamlX.Transform.XamlTransformHelpers.TryConvertValue(AstTransformationContext context, IXamlAstValueNode node, IXamlType type, XamlAstClrProperty propertyContext, IXamlAstValueNode& rv)
   at XamlX.Transform.Transformers.ConstructableObjectTransformer.TransformArgumentsAndGetConstructor(AstTransformationContext context, XamlAstObjectNode n)
   at XamlX.Transform.Transformers.ConstructableObjectTransformer.Transform(AstTransformationContext context, IXamlAstNode node) in /_/src/Markup/Avalonia.Markup.Xaml.Loader/xamlil.github/src/XamlX/Transform/Transformers/ConstructableObjectTransformer.cs:line 56
   at XamlX.Transform.AstTransformationContext.Visitor.Visit(IXamlAstNode node) in /_/src/Markup/Avalonia.Markup.Xaml.Loader/xamlil.github/src/XamlX/Transform/AstTransformationContext.cs:line 62 Line 24, position 9.' Line number '24' and line position '9'.

Am I missing an assembly? I'm using Avalonia 0.10.13.
Using this icons: https://github.com/AvaloniaUtils/Material.Icons.Avalonia

Kindest regards - StanDeMan

**Reply by timunie** [reply]
I have no PC atm, so I cant provide a running sample. Maybe this from the docs can help you

https://docs.avaloniaui.net/docs/controls/togglebutton

Maybe the below code also works, but as I said I can't test it today.

```xml
<UserControl 
	xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	xmlns:icons="using:Material.Icons.Avalonia"
    mc:Ignorable="d" 
	d:DesignWidth="24" 
	d:DesignHeight="24"
    x:Class="WhatEver.Gui.UserInterface.Controls.MinMaxToggelButton">
	<StackPanel>
		<Border
			Background="Transparent">
			<ToggleButton 
				Width="24"
				Height="24"
				IsChecked="True">
				<ToggleButton.Template>
					<ControlTemplate>
						<Panel>
							<icons:MaterialIcon
								Width="20"
								Height="20"
								IsVisible="{Binding !IsChecked, RelativeSource={RelativeSource TemplatedParent}}"
								Kind="ArrowCollapseAll"/>
						<icons:MaterialIcon 
								Width="20"
								Height="20"	
								IsVisible="{Binding IsChecked, RelativeSource={RelativeSource TemplatedParent}}"
								Kind="ArrowExpandAll"/>
						</Panel>
					</ControlTemplate>	
				</ToggleButton.Template>
			</ToggleButton>  
		</Border>
	</StackPanel>
</UserControl>
```

Edit: maybe TemplateBinding does not work with !

**Reply by StanDeMan** [reply]
Thanks a lot for your help. That was it. It's working and it's so simple. 
Maybe it saves someone from wasting a lot of time.

Kindest regards 
StanDeMan
